debezium-core/src/main/java/io/debezium/relational/Key.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.relational;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.annotation.Immutable;
+import io.debezium.function.Predicates;
+
+/**
+ * An immutable definition of the Table.Key. By default the key will be compound of the primary key column. <br>
+ * Key can be customized with a {@code CustomKeyMapper} 
+ * 
+ * @author Guillaume Rosauro
+ *
+ */
+@Immutable
+public class Key {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(Key.class);
+    
+    private final Table table;
+    private final KeyMapper keyMapper;
+    
+    
+    private Key(Table table, KeyMapper keyMapper) {
+        this.table = table;
+        this.keyMapper = keyMapper;
+    }
+
+    /**
+     * 
+     * @return the columns {@link Column} taking part of the key.
+     */
+    public List<Column> keyColumns() {
+        return keyMapper.matches(table);
+    }
+    
+    public static class Builder {
+        
+        private final Table table;
+        private KeyMapper keyMapper = IdentityKeyMapper.getInstance();
+        
+        public Builder( Table table) {
+            this.table = table;
+        }
+        
+        public Builder customKeyMapper(CustomKeyMapper customKeyMapper) {
+            if (customKeyMapper != null) {
+                this.keyMapper = customKeyMapper;
+            }
+            return this;
+        }
+        
+        public Key build() {
+            return new Key(table, keyMapper);
+        }
+        
+    }
+    
+    @FunctionalInterface
+    public static interface KeyMapper {
+        
+        /**
+         * 
+         * @param table {@code Table} 
+         * @return the list {@code Column} describing the key.
+         */
+        List<Column> matches(Table table);
+        
+    }
+    
+    /**
+     * 
+     * Default Key mapper using PK as key.
+     *
+     */
+    @FunctionalInterface
+    private static interface IdentityKeyMapper extends KeyMapper {
+        
+        public static IdentityKeyMapper getInstance() {
+            return (table) -> {
+                return table.columns()
+                        .stream()
+                        .filter(c -> table.isPrimaryKeyColumn(c.name()))
```

Comment:
Here's something I just realize: we might end up with the wrong key column order, in case key columns are in a different order than in the table itself. Say I have this table definition:

```
CREATE TABLE INFO (
    t1ID INT,
    t2ID INT,
    PRIMARY KEY (t2ID,t1ID)
) 
```

Then I think we'd end up with t1ID,t2ID instead of t2ID,t1ID as the Kafka message key. That's independent of this change, though, and was the case before, too. Something we should fix eventually, but it can be done separately, as it's not a regression. @jpechane, @Naros, WDYT?

Reply:
Logged https://issues.jboss.org/browse/DBZ-1507 for this.